### PR TITLE
[compiler] Improve conversion from PyTorch IR to Relay

### PR DESCRIPTION
**Pull request by bwasti**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24 [WIP] Use PackedFunc to enable custom TVM operator registration from python
* #23 [benchmark] Add initial benchmark file
* **#22 [compiler] Improve conversion from PyTorch IR to Relay**
* #21 [testing] Improve testing to include models, image inputs and autotuned parameters
* #17 [compiler] Add None type handling with sentinel value
* #14 [compiler] Support operators with multiple outputs

### CODE REVIEW

**Comment by bwasti** [comment]
This PR predominantly fixes bugs in conversion

File: torch_tvm/operators.cpp [hunk 1]
```diff
@@ -186,6 +217,25 @@ RegisterTVMOperator reg({
        auto out = tvm::relay::CallNode::make(op, inputs, tvm::Attrs(), {});
        return out;
      }},
+    {Symbol::fromQualString("aten::threshold_"),
```
Comment (wanchaol) [inline torch_tvm/operators.cpp]: are you getting `aten::threshold_` from the trace rather than `aten::relu`?

File: torch_tvm/version.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# This file is generated by setup.py. DO NOT EDIT!
```
Comment (wanchaol) [inline torch_tvm/version.py]: what's this used for? can we add it to gitignore instead?

File: torch_tvm/operators.cpp [hunk 3]
```diff
@@ -186,6 +217,25 @@ RegisterTVMOperator reg({
...
+     [](Node* node, tvm::Array<tvm::relay::Expr> inputs) {
+       AT_CHECK(!relayIsNone(inputs[0]));
+       AT_CHECK(!relayIsNone(inputs[1]));
+       AT_CHECK(!relayIsNone(inputs[2]));
+       auto d = relayToConstant<float>(inputs[1]);
+       AT_CHECK(d <  1e-7, "aten::threshold_ only supported for threshold 0, got", d);
+       AT_CHECK(d > -1e-7, "aten::threshold_ only supported for threshold 0, got", d);
+       AT_CHECK(d <  1e-7, "aten::threshold_ only supported for threshold 0, got", d);
```
Comment (wanchaol) [inline torch_tvm/operators.cpp]: it looks like there are 2 duplicate checks here

File: torch_tvm/operators.cpp [hunk 4]
```diff
@@ -186,6 +217,25 @@ RegisterTVMOperator reg({
...
+       AT_CHECK(d > -1e-7, "aten::threshold_ only supported for threshold 0, got", d);
+       d = relayToConstant<float>(inputs[2]);
+       AT_CHECK(d <  1e-7, "aten::threshold_ only supported for value 0, got", d);
+       AT_CHECK(d > -1e-7, "aten::threshold_ only supported for value 0, got", d);
+       AT_CHECK(d <  1e-7, "aten::threshold_ only supported for value 0, got", d);
```
Comment (wanchaol) [inline torch_tvm/operators.cpp]: same here

File: torch_tvm/operators.cpp (hunk 1 above)
Comment (wanchaol) [reply torch_tvm/operators.cpp]: `aten::threshold` is too generic to pass to tvm, I think we should only consider `aten::relu`

File: torch_tvm/compiler.cpp [hunk 5]
```diff
@@ -23,24 +28,33 @@ tvm::relay::Var TVMCompiler::convertToRelay(const Value* val) {
 }
 
 tvm::relay::Expr TVMCompiler::convertToRelay(const IValue& val) {
+  // All doubles are converted to floats
```
Comment (wanchaol) [inline torch_tvm/compiler.cpp]: looks like we can convert to `float64` for double, can you add a comment here why the conversion to `float32` is necessary?

File: torch_tvm/compiler.cpp [hunk 6]
```diff
@@ -69,14 +85,15 @@ tvm::relay::Expr TVMCompiler::convertToRelay(const IValue& val) {
 }
 
 tvm::relay::Function TVMCompiler::convertToRelay(
-    std::shared_ptr<Graph> subgraph) {
+    std::shared_ptr<Graph> subgraph, std::vector<Value*>* input_values) {
```
Comment (wanchaol) [inline torch_tvm/compiler.cpp]: just take a reference of the vector as argument?

File: torch_tvm/compiler.cpp [hunk 7]
```diff
@@ -199,11 +224,21 @@ void TVMCompiler::run(Stack& stack) {
...
+  for (auto i = 0; i < input_values_.size(); ++i) {//subgraph_->inputs().size(); ++i) {
+    //auto* ivalue = value_to_ivalue[subgraph_->inputs()[i]];
+    auto* value = input_values_[i];
+    if (!value_to_ivalue.count(value)) {
+      auto optional_ivalue = toIValue(value);
+      AT_ASSERT(optional_ivalue.has_value());
+      value_to_ivalue[value] = optional_ivalue.value();
+    }
+    auto ivalue = value_to_ivalue.at(input_values_[i]);
+    auto tensor = ivalue.toTensor().to(caffe2::TypeMeta::Make<float>());
```
Comment (wanchaol) [inline torch_tvm/compiler.cpp]: I think we can do something like `ivalue.toTensor().to(at::kFloat)`

File: torch_tvm/version.py (hunk 2 above)
Comment (bwasti) [reply torch_tvm/version.py]: ah this shouldn't even be in this commit

File: torch_tvm/operators.cpp (hunk 1 above)
Comment (bwasti) [reply torch_tvm/operators.cpp]: I'm getting threshold_ from the trace, not relu

File: torch_tvm/compiler.cpp (hunk 5 above)
Comment (bwasti) [reply torch_tvm/compiler.cpp]: it's not really necessary, I'll swap to fp64

File: torch_tvm/compiler.cpp (hunk 6 above)
Comment (bwasti) [reply torch_tvm/compiler.cpp]: input_values is being written to, so I'm passing it as a pointer

**Pull request closed without merging**